### PR TITLE
[FIX] web : save chatbot lines without error

### PR DESCRIPTION
**Pull request by Mtaylorr**
Issue
======
When you update a chatbot a line and then save the chatbot form it will throw an error.

Steps to reproduce the error:
=============================
1-Go to Live Chat/configuration/Chatbots and select one 
2-update one line in the script (add some letters in one name) 
3-save the chatbot form
4- we have this error : 
```
TypeError: Cannot read properties of undefined (reading 'type')
    at Record.__syncData (http://localhost:8069/web/assets/1916-11953ed/web.assets_backend.min.js:2543:205)
    at Record.setup (http://localhost:8069/web/assets/1916-11953ed/web.assets_backend.min.js:2505:92)
    at new DataPoint (http://localhost:8069/web/assets/1916-11953ed/web.assets_backend.min.js:2496:6)
    at new Record (http://localhost:8069/web/assets/1916-11953ed/web.assets_backend.min.js:2503:31)
    at http://localhost:8069/web/assets/1916-11953ed/web.assets_backend.min.js:2587:220
    at Array.map (<anonymous>)
    at StaticList.__syncData (http://localhost:8069/web/assets/1916-11953ed/web.assets_backend.min.js:2587:106)
    at StaticList.setup (http://localhost:8069/web/assets/1916-11953ed/web.assets_backend.min.js:2586:52)
    at new DataPoint (http://localhost:8069/web/assets/1916-11953ed/web.assets_backend.min.js:2496:6)
    at new StaticList (http://localhost:8069/web/assets/1916-11953ed/web.assets_backend.min.js:2584:39)
```

Problem:
========
Some fieldName doesn't exist in the list of fields.

Solution:
=========
Swapped the values in the assign in duplicateDataPoint.

opw-3454514

### CODE REVIEW

**Review by nle-odoo — APPROVED** [review]
seems to work for the given case

but usually we would rather fix the origin of the issue (why would the field not be defined here?). For the current fix situation seems ok (because we don't do anything special to an integer), but for other type of field there could still be an issue.

Here it seems there is a conflict between:

- the many2many_tags used in the list view to show answer_ids field in the chatbox.script.step list view
- the list view that is used to show answer_ids in the chatbox.script.step form view and contains sequence:

```xml
<tree editable="bottom">
    <field name="sequence" widget="handle"/>
    <field name="display_name" invisible="1"/>
    <field name="name"/>
    <field name="redirect_link" string="Optional Link"/>
</tree>
```
It seemed to me that the field in the multiple views for the same element should have been merged and so that there would not be any issue of a field missing.

But I don't know what the best solution would be (and this situation of having a many2many_tags and list view with more fields should be present in a lot of other views, so I'm not sure why this is failing here =/)

**Review by hubvd — COMMENTED** [review]
The actual cause comes from this commit:

https://github.com/odoo/odoo/commit/f4f0f783183507df8227b37fe1234c256325df6d#diff-415538caff9e7dc43b4a5667f0218322046ab26a84e23094a44a1ab715243791

Where they do

```diff
diff --git a/addons/web/static/src/views/basic_relational_model.js b/addons/web/static/src/views/basic_relational_model.js
index 808ab4295a8f..3e1b773002d2 100644
--- a/addons/web/static/src/views/basic_relational_model.js
+++ b/addons/web/static/src/views/basic_relational_model.js
@@ -1219,9 +1219,9 @@ export class RelationalModel extends Model {
                             const fieldViews = fieldInfo.views || fieldInfo.fieldsInfo || {};
                             const defaultFieldInfo = legRec.data[name].fieldsInfo.default;
                             Object.values(fieldViews).forEach((fieldView) => {
-                                _.defaults(fieldView.fields, defaultFieldInfo);
+                                Object.assign(defaultFieldInfo, fieldView.fields);
                                 Object.values(fieldView.fieldsInfo).forEach((x2mFieldInfo) => {
-                                    _.defaults(x2mFieldInfo, defaultFieldInfo);
+                                    Object.assign(defaultFieldInfo, x2mFieldInfo);
                                 });
                             });
                         }

```

This mutates `defaultFieldInfo` instead of `x2mFieldInfo & fieldView.fieldsInfo`.
You can simply swap both parameters.

```diff
diff --git a/addons/web/static/src/views/basic_relational_model.js b/addons/web/static/src/views/basic_relational_model.js
index 597239827ebc..8f6e071f4394 100644
--- a/addons/web/static/src/views/basic_relational_model.js
+++ b/addons/web/static/src/views/basic_relational_model.js
@@ -1226,9 +1226,9 @@ export class RelationalModel extends Model {
                             const fieldViews = fieldInfo.views || fieldInfo.fieldsInfo || {};
                             const defaultFieldInfo = legRec.data[name].fieldsInfo.default;
                             Object.values(fieldViews).forEach((fieldView) => {
-                                Object.assign(defaultFieldInfo, fieldView.fields);
+                                Object.assign(fieldView.fields, defaultFieldInfo);
                                 Object.values(fieldView.fieldsInfo).forEach((x2mFieldInfo) => {
-                                    Object.assign(defaultFieldInfo, x2mFieldInfo);
+                                    Object.assign(x2mFieldInfo, defaultFieldInfo);
                                 });
                             });
                         }
```

**Review by hubvd — APPROVED** [review]

File: addons/web/static/tests/views/form/form_view_tests.js [hunk 1]
```diff
@@ -13322,4 +13322,75 @@ QUnit.module("Views", (hooks) => {
...
+        await click(target.querySelectorAll(".o_data_cell")[0]);
+        assert.hasClass(
+            document.querySelector("body"),
+            "modal-open",
+            "body should have modal-open class"
+        );
+        target.querySelector(".modal-body [name='display_name'] input").value = "updated";
+        await click(target.querySelector(".modal-footer .o_form_button_save"));
+        assert.containsOnce(target, ".o_form_button_save");
+        await click(target.querySelector(".o_form_button_save"));
```
Comment (FrancoisGe) [inline addons/web/static/tests/views/form/form_view_tests.js:13393]: add spacing maybe after block of asserts

File: addons/web/static/tests/views/form/form_view_tests.js [hunk 2]
```diff
@@ -13322,4 +13322,75 @@ QUnit.module("Views", (hooks) => {
...
+
+        serverData.models.extra = {
+            fields: {
+                name: { string: "Char", type: "char" },
+            },
+            records: [
+                { id: 1, display_name: "ma", name: "ma" },
+                { id: 2, display_name: "cr", name: "cr" },
+            ],
+        };
```
Comment (FrancoisGe) [inline addons/web/static/tests/views/form/form_view_tests.js:13358]: Why don t use the default serverData ?

Comment (Mtaylorr) [reply addons/web/static/tests/views/form/form_view_tests.js]: I changed it to make use more of default serverData , but had to change some models to get the intended scenario.

**Comment by Mtaylorr** [comment]
@FrancoisGe can you approve this FIx if everything is ok for you ?

**Comment by FrancoisGe** [comment]
@robodoo r+

**Pull request closed without merging**